### PR TITLE
feat: `bring dynamodb`

### PR DESCRIPTION
**Pull request by skyrpex**
- Interface similar to the original DynamoDB SDK
- Includes streams support
- Uses one container for all tables (sim)

Closes https://github.com/winglang/wing/issues/4949

### CODE REVIEW

**Comment by ekeren** [comment]
@skyrpex what is the status of this?

**Comment by skyrpex** [comment]
Hmmm not much I think! 

- Extract interface & abstract implementation from current table class
- Implement tf-aws table
- Marshall dynamodb stream event
- Not sure if we're missing critical type defnitions or methods

**Comment by ekeren** [comment]
Sounds promising, does it also work locally?

**Comment by skyrpex** [comment]
> Sounds promising, does it also work locally?

Yes, it works locally :)

File: dynamodb/dynamodb.mjs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      shardIterator = recordsData.NextShardIterator;
+
+      // AWS DynamoDB fetches records 4 times a second.
+      await new Promise((resolve) => setTimeout(resolve, 250));
+    }
+  }
+};
+
+const processRecords = async (endpoint, tableName, handler) => {
+  while (true) {
```
Comment (eladb) [inline dynamodb/dynamodb.mjs:109]: How do we break this loop on shutdown?

File: dynamodb/dynamodb.mjs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  }
+};
+
+export const processRecordsAsync = async (endpoint, tableName, handler) => {
+  processRecords(endpoint, tableName, handler).catch((error) => {
+    if (error.message.includes("ECONNREFUSED")) {
+      // Ignore. This error happens when reloading the console.
+      // We can safely end the execution here.
+    } else {
+      console.error(error);
```
Comment (eladb) [inline dynamodb/dynamodb.mjs:149]: No need to throw?

File: dynamodb/dynamodb.sim.w [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+bring util;
+bring cloud;
+bring "./dynamodb-types.w" as dynamodb_types;
+bring "./dynamodb-client.w" as dynamodb_client;
+
+interface Process {
+  inflight kill(): void;
+}
+
+struct SpawnOptions {
```
Comment (eladb) [inline dynamodb/dynamodb.sim.w:11]: @Chriscbr any idea when this will be supported in the sdk? This is the 4th implementation I've seen of this code (and all of them have issues I am sure)

File: dynamodb/dynamodb.sim.w [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  new() {
+    let containerName = "winglibs-dynamodb-{util.uuidv4()}";
+
+    let state = new sim.State();
+    this.endpoint = state.token("endpoint");
+
+    new cloud.Service(inflight () => {
+      let port = Util.getPort();
+
+      let docker = Util.spawn(
```
Comment (eladb) [inline dynamodb/dynamodb.sim.w:48]: Would it make sense to use containers.Workload here?

File: dynamodb/dynamodb.sim.w [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+          }
+        },
+      );
+
+      return () => {
+        docker.kill();
+      };
+    });
+
+    // The host will be ready when the endpoint is set.
```
Comment (eladb) [inline dynamodb/dynamodb.sim.w:71]: Why is this needed?

File: dynamodb/dynamodb.sim.w [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  tableName: str;
+
+  new(props: dynamodb_types.TableProps) {
+    this.host = Host.of(this);
+
+    let tableName = this.node.addr;
+    let state = new sim.State();
+    this.tableName = state.token("tableName");
+
+    new cloud.Service(inflight () => {
```
Comment (eladb) [inline dynamodb/dynamodb.sim.w:102]: @Chriscbr please take a look at this- another use case for `sim.Resource` we can use to validate the APIs.

File: dynamodb/dynamodb.sim.w [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  pub connection(): dynamodb_types.Connection {
+    return {
+      endpoint: this.host.endpoint,
+      tableName: this.tableName,
+    };
+  }
+
+  pub setStreamConsumer(handler: inflight (dynamodb_types.StreamRecord): void) {
```
Comment (eladb) [inline dynamodb/dynamodb.sim.w:171]: Need some shutdown code here

**Review by eladb — CHANGES_REQUESTED** [review]
Crazy project!

File: dynamodb/dynamodb.sim.w [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        });
+      }
+
+      util.waitUntil(() => {
+        try {
+          client.createTable({
+            TableName: tableName,
+            AttributeDefinitions: attributeDefinitions.copy(),
+            KeySchema: keySchemas.copy(),
+            BillingMode: "PAY_PER_REQUEST",
```
Comment (staycoolcall911) [inline dynamodb/dynamodb.sim.w:138]: Should we add `pointInTimeRecovery`? https://github.com/winglang/wing/pull/5730

File: dynamodb/dynamodb-client.w [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    this.tableName = props.tableName;
+    this.client = Util.createDocumentClient({
+      region: props.region,
+      credentials: props.credentials,
+      endpoint: props.endpoint,
+    });
+  }
+
+  pub inflight delete(options: dynamodb_types.DeleteOptions): dynamodb_types.DeleteOutput {
+    let response = this.client.delete({
```
[the remark this answers is not in the document]
Comment (eladb) [reply dynamodb/dynamodb-client.w:52]: let's do it and see how it feels.

File: dynamodb/dynamodb-client.w [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+  endpoint: str?;
+  region: str?;
+  credentials: Credentials?;
+}
+
+pub inflight class Client impl dynamodb_types.IClient {
+  inflight tableName: str;
+  inflight client: DocumentClient;
+
+  inflight new(props: ClientProps) {
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply dynamodb/dynamodb-client.w:42]: The props must include the tableName, so aren't optional.

File: dynamodb/dynamodb-client.w [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+  inflight scan(input: Json): Json;
+  inflight transactGet(input: Json): Json;
+  inflight transactWrite(input: Json): Json;
+  inflight update(input: Json): Json;
+}
+
+struct Credentials {
+  accessKeyId: str;
+  secretAccessKey: str;
+}
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply dynamodb/dynamodb-client.w:19]: Yes, we need to pass fake credentials to the DynamoDB container in the sim, otherwise the client will try to gather the credentials from environment variables. For example, this is what I get:

```
The Wing Console is running at http://localhost:3000/
Error: The SSO session associated with this profile has expired. To refresh this SSO session run aws sso login with the corresponding profile.
   --> dynamodb-client.w:66:12
```

File: dynamodb/README.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    {
+      name: "id",
+      type: "S",
+    },
+  ],
+  hashKey: "id",
+);
+
+// Streams.
+table.setStreamConsumer(inflight (record) => {
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply dynamodb/README.md:29]: Maybe. It'd make the implementation more complicated, though.

IMO we should stick to this raw implementation, and see if users actually need more advanced patterns.

[the remark this answers is not in the document]
Comment (skyrpex) [reply dynamodb/README.md:29]: DynamoDB always emits events for all records. Not sure if that's what you were asking.

File: dynamodb/dynamodb.sim.w (hunk 4 above)
Comment (skyrpex) [reply dynamodb/dynamodb.sim.w:48]: I guess, I just tried it and the resulting code is much cleaner, but I noticed that the containers doesn't get removed when I Ctrl+C the `wing it` process.

Maybe we can improve the containers winglib. The difference is that I spawn the docker container without detach and with the `--rm` option.

File: dynamodb/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (skyrpex) [reply dynamodb/README.md:29]: In that case, it would need to be implemented manually for the sim AFAIK. Maybe we could implement it at a later stage, with proper testing.

File: dynamodb/dynamodb.sim.w (hunk 5 above)
Comment (skyrpex) [reply dynamodb/dynamodb.sim.w:71]: I got BSODs about functions not getting resolved otherwise.

File: dynamodb/dynamodb.sim.w (hunk 7 above)
Comment (skyrpex) [reply dynamodb/dynamodb.sim.w]: I haven't encountered a situation where this keeps running. If the table container stops being available, it will end up halting the loop.

**Comment by skyrpex** [comment]
@Chriscbr done! Had to add support for global secondary indexes :)

Didn't include https://github.com/winglang/wing/blob/main/examples/tests/sdk_tests/dynamodb-table/aws-dynamodb.test.w because I didn't see the point of it.

**Comment by Chriscbr** [comment]
Sweet, thanks

File: dynamodb/dynamodb.sim.w (hunk 6 above)
Comment (Chriscbr) [reply dynamodb/dynamodb.sim.w:67]: Thanks for the reference point

File: dynamodb/dynamodb-client.w [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    return unsafeCast(this.client.transactWrite({
+      TransactItems: transactItems.copy(),
+    }));
+  }
+
+  pub inflight scan(options: dynamodb_types.ScanOptions?): dynamodb_types.ScanOutput {
+    let input: MutJson = options ?? {};
+    input.set("TableName", this.tableName);
+    return unsafeCast(this.client.scan(input));
```
Comment (Chriscbr) [inline dynamodb/dynamodb-client.w:108]: Curious for my own understanding, how tricky is it to implement this method without `unsafeCast`?

Is `let input: MutJson = options ?? {}` a workaround for Winglang not allowing you set struct fields?

**Review by Chriscbr — APPROVED** [review]
I'm not intimately familiar with all the API surface area of DynamoDB so my review mostly focused on code quality. I added a few questions but nothing blocking.

File: dynamodb/package.json [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    "@cdktf/provider-aws": "^19.9.0",
+    "@winglibs/containers": "^0.0.22",
+    "get-port": "^7.0.0"
+  },
+  "devDependencies": {
+    "@cdktf/provider-helm": "^10.0.0",
+    "@cdktf/provider-kubernetes": "^11.2.0",
+    "@cdktf/provider-null": "^10.0.0",
+    "cdk8s-plus-27": "^2.7.84"
+  }
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply dynamodb/package.json:29]: Yes, they are needed by the containers winglibs. Otherwise, it will just fail to run (on my machine, at least).

File: dynamodb/dynamodb-client.w (hunk 13 above)
Comment (skyrpex) [reply dynamodb/dynamodb-client.w:108]: > It looks like the loop is broken by throwing an error

It is intended. There's a try catch later that ignore certain kind of errors, but throw otherwise.

> how tricky is it to implement this method without unsafeCast?

I guess we could add the proper return types to `DocumentClient` in `dynamodb/dynamodb-client.w` now that they return the original information, but the options still need the `TableName` to be added manually... Any ideas?

**Review by eladb — APPROVED** [review]
